### PR TITLE
OpenAPI 3.x support

### PR DESCRIPTION
**Pull request by daniel-shuy**
Closes meedan/check#8

I recommend reviewing this PR by commit, as it is quite big.

---

This PR adds OpenAPI 3.x support by replacing [swagger-docs](https://github.com/richhollis/swagger-docs) (which is still stuck on Swagger 1.x) with [Rswag](https://github.com/rswag/rswag).

Note that the Swagger UI URL is now at `/api-docs` instead of `/api`.

---

Unlike `swagger-docs`, `Rswag` embeds Swagger UI, and exposes a middleware endpoint for it (defined in [config/routes.rb](https://github.com/meedan/pender/pull/233/commits/8860091f11da977c02c3ae9dba269c6354d6d858#diff-959bc9abc46a55332bb64d5155a79323afa75a50ec1a2137ddd22d926f62c6c5)). This ensures that the Swagger UI is automatically updated with the `Rswag` gem.

Since the Swagger UI is up-to-date, the UI is much better:
image
image
image

---

`Rswag` generates OpenAPI documentation from [rspec-rails](https://github.com/rspec/rspec-rails) tests. I've placed them in `spec/requests/api`, which is the default `rails generate rspec:swagger` location. See:
- [spec/requests/api/base_api_spec.rb](https://github.com/meedan/pender/pull/233/commits/aee73b00e1f19f41e6c6db5db127fadb4a99ad1d#diff-67cf9b88436474d7327263c179301d48f2f6cba0218053690f30d5b3ef95899d)
- [spec/requests/api/medias_spec.rb](https://github.com/meedan/pender/pull/233/commits/e0e8cb65195ec747a869c49dfed700326b8379b1#diff-b2c51c61291a55b9a9e7673b205ebc31d12b9217687734e377fb6394d33cbdfc)

To generate the OpenAPI documentation, run:
```
RAILS_ENV=test rake rswag
```

I have also [included it as part of the `lapis:docs` task](https://github.com/meedan/pender/pull/233/commits/8860091f11da977c02c3ae9dba269c6354d6d858#diff-f48ddfd1eec38c4d39f84e195259371f8397cd30146f46b113bdcb4590ed262c).

The OpenAPI documentation is generated at `swagger/v1/swagger.yaml`. I have excluded it from [.gitignore](https://github.com/meedan/pender/pull/233/commits/eaf3effb736da704dae3a648f601d3262b77bec3). Is this ok? What's the release/deployment process?

---

Rswag has the ability to [automatically generate examples using responses from tests](https://github.com/rswag/rswag#enable-auto-generation-examples-from-responses):
image
For convenience, [I have created a RSpec shared context](https://github.com/meedan/pender/pull/233/commits/2150861480bfa42d7ac7c0a74b515c51a434d7a2#diff-1e8cf3de001b876faaf2655349c90959c6a725c8ead4744a692aa285e986a4ff) for this. To use it, simply put:
```ruby
include_context 'generate examples'
```
at the end of whichever test that you wish to automatically generate examples for.

Enabling `Rswag`'s [auto generation examples from responses](https://github.com/meedan/pender/pull/233/commits/2150861480bfa42d7ac7c0a74b515c51a434d7a2) requires disabling `Rswag`'s `--dry-run` flag. This flag causes `Rswag` to skip the tests when generating the documentation, and is enabled by default. Disabling it will require the tests to pass to generate the documentation. If you want to generate the OpenAPI documentation without having to run the tests, you can temporarily re-enable the `--dry-run` flag (do note that however that will disable `auto generation examples from responses`).

Note that there is currently a bug in `Rswag` (rswag/rswag#359) that requires configuring the `--dry-run` flag in `config/application.rb` instead of `config/environment/test.rb` (see https://github.com/meedan/pender/pull/233/commits/e7f370cd66ba590bb1d3938340f64fe5cd49a908).

---

I have also taken the opportunity to add Schema information:
image
image

---

The commit that needs to be reviewed the most is to [Enable SimpleCov for Rswag](https://github.com/meedan/pender/pull/233/commits/8c780d8e56a741ccac40106d557db89ef760a42c). I'm not sure if I'm doing this correctly, especially whether or not the results will be properly merged with the existing integration tests'.

---

Note that this PR contains workarounds for a few bugs in `Rswag`, please remove them once the bugs are fixed:
- rswag/rswag#359 - [workaround](https://github.com/meedan/pender/pull/233/commits/e7f370cd66ba590bb1d3938340f64fe5cd49a908)
- rswag/rswag#325 - [workaround](https://github.com/meedan/pender/pull/233/commits/c764bff5f730c6a5d8afc8cda3e3b73c7b0e0e51)
- rswag/rswag#317 - [workaround](https://github.com/meedan/pender/pull/233/commits/f5dfc70afbaf98f4fda99cd02bab1f3a8414fbfb)

The workaround for rswag/rswag#317 is the most significant, as it requires using the [jetpackworkflow/rswag](https://github.com/jetpackworkflow/rswag) fork, which means that you will not be able to receive mainline updates in the meantime, unless you create your own `Rswag` fork.

---

As part of removing `swagger-docs`, I had to remove `swagger-docs`' Swagger JSON to Markdown functionality, which was used to generate [doc/api.md](https://github.com/daniel-shuy/pender/blob/feature/openapi/doc/api.md). I couldn't find any Ruby gem that could convert OpenAPI 3.x YAML/JSON to Markdown.

However, comparing [doc/api.md](https://github.com/daniel-shuy/pender/blob/feature/openapi/doc/api.md) to [public/api-docs.json](https://github.com/meedan/pender/pull/233/commits/193a80b4563241ea3daac1cf33b5d25bf70958ea#diff-d357a6c16b9b96829d068bce1eeb2607ea404bd2ba17aefa5084155bf1888899), it seems like they are no longer in line anyways, therefore I removed the [`api.md` `Makefile` task](https://github.com/meedan/pender/pull/233/commits/193a80b4563241ea3daac1cf33b5d25bf70958ea#diff-f48ddfd1eec38c4d39f84e195259371f8397cd30146f46b113bdcb4590ed262c), but kept the [doc/api.md](https://github.com/daniel-shuy/pender/blob/feature/openapi/doc/api.md) file.

Alternately, we can use [widdershins](https://github.com/Mermade/widdershins) to convert the OpenAPI 3.x YAML to Markdown, and update the new OpenAPI definition to match the current [doc/api.md](https://github.com/daniel-shuy/pender/blob/feature/openapi/doc/api.md). This however requires `npm`.

### CODE REVIEW

**Comment by danielafeitosa** [comment]
Hi @daniel-shuy !
Sorry for taking so long to review your PR. 
The tests are running here, we'll see if your PR affected the tests: https://travis-ci.com/github/meedan/pender/builds/195310578

I loved how you documented all the changes you made!

> The OpenAPI documentation is generated at swagger/v1/swagger.yaml. I have excluded it from .gitignore. Is this ok? What's the release/deployment process?

You should remove from `.gitignore`. We prefer to have the file already in the repository so the developers don't need to generate it. The file only need to be generated when something changes on API

> However, comparing doc/api.md to public/api-docs.json, it seems like they are no longer in line anyways, therefore I removed the api.md Makefile task, but kept the doc/api.md file.

It's great!

> Alternately, we can use widdershins to convert the OpenAPI 3.x YAML to Markdown, and update the new OpenAPI definition to match the current doc/api.md. This however requires npm.

It doesn't need to be done on this PR. But if you want to create a new PR converting it with `npm`, it will be a nice contribution.

**Comment by daniel-shuy** [comment]
> You should remove from .gitignore. We prefer to have the file already in the repository so the developers don't need to generate it. The file only need to be generated when something changes on API

Alright, I've removed it

File: test/controllers/base_api_controller_test.rb [hunk 1]
```diff
@@ -6,28 +6,6 @@ def setup
...
-    get :about, format: :json
-    assert_response :success
-    response = JSON.parse(@response.body)
-    assert_equal ['data', 'type'], response.keys.sort
-    assert_equal ['archivers', 'name', 'version'], response['data'].keys.sort
-    assert_equal 'about', response['type']
-    assert_equal 'Keep', response['data']['name']
-    assert_equal VERSION, response['data']['version']
-  end
-
```
Comment (danielafeitosa) [inline test/controllers/base_api_controller_test.rb:30]: @daniel-shuy why did you remove these tests? They were not affected by your changes, they are still passing.
Could you add them back, please?

File: test/controllers/medias_controller_test.rb [hunk 2]
```diff
@@ -7,23 +7,6 @@ def setup
...
-    get :index, url: 'http://meedan.com', format: :json
-    assert_response 401
-  end
-
-  test "should parse media" do
-    authenticate_with_token
-    get :index, url: 'http://twitter.com/meedan', format: :json
-    assert_response :success
-  end
-
```
Comment (danielafeitosa) [inline test/controllers/medias_controller_test.rb:26]: @daniel-shuy why did you remove these tests? They were not affected by your changes, they are also passing.
Could you add them back, please?

File: test/controllers/medias_controller_test.rb [hunk 3]
```diff
@@ -255,40 +238,13 @@ def setup
...
-    authenticate_with_token
-    get :index, url: 'https://twitter.com/anxiaostudio', format: :json
-    assert_response 429
-    assert_equal 123, JSON.parse(@response.body)['data']['message']
-
-    Twitter::REST::Client.any_instance.unstub(:user)
-    Twitter::Error::TooManyRequests.any_instance.unstub(:rate_limit)
-  end
-
   test "should render custom HTML if provided by parser" do
```
Comment (danielafeitosa) [inline test/controllers/medias_controller_test.rb:241]: @daniel-shuy why did you remove these tests? They were not affected by your changes, they are also passing.
Could you add them back, please?

File: test/controllers/medias_controller_test.rb [hunk 4]
```diff
@@ -255,40 +238,13 @@ def setup
...
-    assert_equal 'The URL is not valid', JSON.parse(@response.body)['data']['message']
-  end
-
-  test "should show error message if url not found" do
-    authenticate_with_token
-    get :index, url: 'http://not-valid', format: :json
-    assert_response 400
-    assert_match /The URL is not valid/, JSON.parse(@response.body)['data']['message']
-  end
-
```
Comment (danielafeitosa) [inline test/controllers/medias_controller_test.rb:291]: @daniel-shuy why did you remove these tests? They were not affected by your changes, they are also passing.
Could you add them back, please?

File: test/controllers/medias_controller_test.rb [hunk 5]
```diff
@@ -376,11 +332,6 @@ def setup
     end
   end
 
-  test "should not clear cache if not authenticated" do
-    delete :delete, url: 'http://test.com', format: 'json'
-    assert_response 401
-  end
-
```
Comment (danielafeitosa) [inline test/controllers/medias_controller_test.rb:383]: @daniel-shuy why did you remove these tests? They were not affected by your changes, they are also passing.
Could you add them back, please?

**Review by danielafeitosa — CHANGES_REQUESTED** [review]
@daniel-shuy your PR is great, thank you!
As soon as you add back the tests from `test/controllers/base_api_controller_test.rb` and `test/controllers/medias_controller_test.rb` I'll merge it

File: test/controllers/base_api_controller_test.rb (hunk 1 above)
Comment (daniel-shuy) [reply test/controllers/base_api_controller_test.rb:30]: I've moved them to the RSpec Rails test (see https://github.com/meedan/pender/pull/233/files#diff-67cf9b88436474d7327263c179301d48f2f6cba0218053690f30d5b3ef95899d)

File: test/controllers/medias_controller_test.rb (hunk 2 above)
Comment (daniel-shuy) [reply test/controllers/medias_controller_test.rb:26]: I've moved them to the RSpec Rails test (see https://github.com/meedan/pender/pull/233/files#diff-b2c51c61291a55b9a9e7673b205ebc31d12b9217687734e377fb6394d33cbdfc)

**Review by danielafeitosa — APPROVED** [review]
